### PR TITLE
Implement type conversion for AstarteDataType

### LINKED ISSUE #10: Implement Type conversion for AstarteDataType (proto)

### PR DESCRIPTION
**Pull request by harlem88**
- Implements type conversion from all scalar and vector Rust type defined in the `astarte_type` proto file.
- Add tests on AstarteDataType conversion trait.
- Use plural names for repeated fields.

Closes #10.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+
+macro_rules! impl_type_conversion_traits {
+    ( {$( ($typ:ty, $astartedatatype:ident) ,)*}) => {
+
+        $(
+               impl From<$typ> for AstarteDataTypeIndividual {
+                    fn from(d: $typ) -> Self {
+                        AstarteDataTypeIndividual {
+                            individual_data: Some(IndividualData::$astartedatatype(d.into())),
+                            }
```
Comment (rbino) [inline src/types.rs:42]: Is this formatted? I'd expect this closed brace to be two spaces on the left

File: src/types.rs [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+            individual_data: Some(IndividualData::AstarteDateTime(system_time.into())),
+        }
+    }
+}
+
+impl From<Vec<f64>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<f64>) -> Self {
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteDoubleArray(AstarteDoubleArray {
+                astarte_double: value,
```
Comment (rbino) [inline src/types.rs:81]: This is more of a "retroactive" review, but now that I see the usage I find the fact that, e.g., `AstarteDoubleArray` has a field called that is called `astarte_double` strange.
First, in the [Protobuf Language Guide](https://developers.google.com/protocol-buffers/docs/proto3) repeated fields always use plural names (see e.g. [here](https://developers.google.com/protocol-buffers/docs/proto3): `results`, `snippets`). Second, while using `astarte_x` is useful at the level above to distinguish between the different `oneof` tags, here it's not and I think it introduced unneeded noise.

All in all, I think that all `AstarteSomethingArray` protobufs should just have a `repeated some_type values` field. This makes them more uniform to access, follows the plural name convention and does not introduce unnecessary long names.

File: src/types.rs [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+        let system_time: SystemTime = value.into();
+
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteDateTime(system_time.into())),
+        }
+    }
+}
+
+impl From<Vec<f64>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<f64>) -> Self {
```
Comment (rbino) [inline src/types.rs:78]: On a similar note, I think in Arrays the argument should be called `values` instead of `value`

File: src/types.rs [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+                AstarteBinaryBlobArray {
+                    astarte_binary_blob: value,
+                },
+            )),
+        }
+    }
+}
+
+impl From<Vec<DateTime<Utc>>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<DateTime<Utc>>) -> Self {
```
Comment (rbino) [inline src/types.rs:142]: The line above says this is not covered by test. Given it's the only non-trivial conversion, I'd add a test for it

**Review by rbino — CHANGES_REQUESTED** [review]

**Review by rbino — APPROVED** [review]

**Review by FrancescoVaiani — APPROVED** [review]

**Review by bettio — APPROVED** [review]

**Pull request merged**